client_new/src/types/client.rs

```diff
@@ -0,0 +1,207 @@
+// Copyright 2020-2022 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+use std::{
+    error::Error,
+    sync::{Arc, RwLock},
+};
+
+use engine::{
+    new_runtime::memories::buffer::Buffer,
+    vault::{view::Record, BoxProvider, ClientId, DbView, RecordHint, VaultId},
+};
+
+use crate::{
+    derive_vault_id,
+    procedures::{
+        FatalProcedureError, Procedure, ProcedureError, ProcedureOutput, Products, Runner, StrongholdProcedure,
+    },
+    ClientError, ClientState, ClientVault, KeyStore, Location, Provider, RecordError, Store,
+};
+
+use super::snapshot;
+
+pub struct Client {
+    // A keystore
+    pub(crate) keystore: Arc<RwLock<KeyStore<Provider>>>,
+
+    // A view on the vault entries
+    pub(crate) db: Arc<RwLock<DbView<Provider>>>,
+
+    // The id of this client
+    pub id: ClientId,
+
+    // Contains the Record Ids for the most recent Record in each vault.
+    pub store: Store,
+}
+
+impl Default for Client {
+    fn default() -> Self {
+        Self {
+            keystore: Arc::new(RwLock::new(KeyStore::default())),
+            db: Arc::new(RwLock::new(DbView::new())),
+            id: ClientId::default(),
+            store: Store::default(),
+        }
+    }
+}
+
+impl Drop for Client {
+    fn drop(&mut self) {}
+}
+
+impl Client {
+    /// Returns [`Self`] with atomic references to the same client to be shared in concurrent setups
+    ///
+    /// # Example
+    /// ```no_run
+    /// ```
+    pub(crate) fn atomic_ref(&self) -> Client {
+        Self {
+            keystore: self.keystore.clone(),
+            db: self.db.clone(),
+            id: self.id,
+            store: self.store.atomic_ref(),
+        }
+    }
+
+    /// Returns an atomic reference to the [`Store`]
+    ///
+    /// # Example
+    /// ```
+    /// ```
+    pub async fn store(&self) -> Store {
+        self.store.atomic_ref()
+    }
+
+    /// Returns a [`Vault`] according to path
+    ///
+    /// # Example
+    /// ```
+    /// ```
+    pub fn vault(&self, vault_path: Location) -> ClientVault {
+        let (vault_id, _) = vault_path.resolve();
+
+        ClientVault {
+            client: self.atomic_ref(),
+            id: vault_id,
+        }
+    }
```

Comment:
With #319 all operations on the vault can be done through procedures. We could consider removing `ClientVault` so that we have a single way of doing things through procedures only.

Reply:
`ClientVault` is a thin wrapper around `Client` and calls all actions directly on `Client`s database.

Reply:
Yes but said actions (`write_secret`, `delete_secret`, `revoke_secret` and `cleanup`) can all be done through procedures (`StrongholdProcedure::WriteVault`, `StrongholdProcedure::RevokeData`, etc). 
Thus the methods are kinda redundant and if we would remove them, we also would not need this `ClientVault` wrapper anymore.

Reply:
I disagree. The `ClientVault` is meant to have as a distinct type to differentiate the actions taken on them. The whole idea of the new client interface is around types.

Reply:
The only point to argue would be to expose this `ClientVault` as an interface on `Client`.

Reply:
My argument was based on what is described in the _Primitive Based Interface_ RFC in #313 in **Reference-level explanation - Vault**: 

> While the vault is a stronghold primitive, it does not have its own type (being the only exception). Most interesting operations on a vault are procedures. Since a client can have multiple vaults, a procedure generally works on the client-level rather than the vault-level (...)

I think it's cleaner to have a unified way of interacting with vaults by always using procedures. It's also easier for the user to understand, rather than having multiple ways of how something can be written to a vault.

But I also see your point. Having a `ClientVault` type is more consistent with how the interface is generally designed.

Anyway, your call 😄.

Reply:
The execution of procedures are on `Client` level, while `ClientVault` offers a more distinct type :D Sure, every action can be done by procedures, but having access to "simpler" function calls might be beneficial, imho.